palette/src/oklab.rs

```diff
@@ -323,7 +608,128 @@ impl_color_div!(Oklab<T>, [l, a, b]);
 impl_array_casts!(Oklab<T>, [T; 3]);
 impl_simd_array_conversion!(Oklab, [l, a, b]);
 
-impl_eq!(Oklab, [l, a, b]);
+impl<T> Oklab<T>
+where
+    T: AbsDiffEq + One + Zero,
+    T::Epsilon: Clone + Real + PartialOrd,
+{
+    /// Returns true, if `lightness == 1`
+    ///
+    /// **Note:** `sRGB` to `Oklab` conversion uses `f32` constants.
+    /// A tolerance `epsilon >= 1e-8` is required to reliably detect white.
+    /// Conversion of `sRGB` via XYZ requires `epsilon >= 1e-5`
+    pub fn is_white(&self, epsilon: T::Epsilon) -> bool {
+        self.l.abs_diff_eq(&T::one(), epsilon)
+    }
+
+    /// Returns true, if `lightness == 0`
+    pub fn is_black(&self, epsilon: T::Epsilon) -> bool {
+        self.l.abs_diff_eq(&T::zero(), epsilon)
+    }
+
+    /// Returns true, if `chroma == 0`
+    pub fn is_grey(&self, epsilon: T::Epsilon) -> bool {
+        self.a.abs_diff_eq(&T::zero(), epsilon.clone()) && self.b.abs_diff_eq(&T::zero(), epsilon)
+    }
+
+    fn both_black_or_both_white(&self, other: &Self, epsilon: T::Epsilon) -> bool {
+        self.is_white(epsilon.clone()) && other.is_white(epsilon.clone())
+            || self.is_black(epsilon.clone()) && other.is_black(epsilon)
+    }
+}
+
+impl<T> PartialEq for Oklab<T>
+where
+    T: PartialEq,
+{
+    /// Returns true, f `l`, `a` and `b` of `self` and `other` are identical.
+    /// This is equality in name only, as with computed floating point numbers there
+    /// always is an error, that must be accounted for, using a tolerance.   
+    fn eq(&self, other: &Self) -> bool {
+        self.l == other.l && self.a == other.a && self.b == other.b
+    }
+}
+impl<T> Eq for Oklab<T> where T: Eq {}
+impl<T> AbsDiffEq for Oklab<T>
+where
+    T: AbsDiffEq + One + Zero,
+    T::Epsilon: Clone + Real + PartialOrd,
+{
+    type Epsilon = T::Epsilon;
+
+    fn default_epsilon() -> Self::Epsilon {
+        T::default_epsilon()
+    }
+
+    /// Returns true, if `self ` and `other` are visually indiscernible, even
+    /// if they hold are both black or both white and their `a` and `b` values differ.
+    ///
+    /// `epsilon` must be large enough to detect white (see [Oklab::is_white])
+    fn abs_diff_eq(&self, other: &Self, epsilon: T::Epsilon) -> bool {
+        self.both_black_or_both_white(other, epsilon.clone())
```

Comment:
From a visual and logical perspective the color angle of black and white colors does not matter. It only matters, if the luminance of the color is changed towards 0.5. However even this may be dangerous, as previous operations, like moving to a different color space and back -- may have corrupted the angle.

Since the color angle does not matter for black or white, the epsilon should not apply to it, imho. However, if you do believe this really should be a plain numerical comparison, is there a trait, that expresses visual / logical equality with or without an epsilon-parameter?

Reply:
You are right that the hue stops mattering when luminance (or saturation) is low, but my main motivation is consistency with other hue based color types. I'm also reluctant to have a hard limit on when it stops mattering, and would rather keep this implementation simple and dumb.

There is currently no trait for this. It may be something to add, but haven't been done yet. But if it's only for the internal logic, it might as well be a regular method for now.

On top of the above, `approx` is no longer a mandatory dependency (made optional in #286) so shouldn't be relied upon outside tests and examples.